Documents ICs from THM

GiudGiud opened the pull request:
and move the more general ones to the framework

this will help me build a "multi-exodus" file restart

GiudGiud committed:
Document SolutionIC, move to framework

GiudGiud committed:
Document ScalarSolutionIC, move to framework

GiudGiud committed:
Document all ICs from the thermal hydraulics module, refs #15968 #19737

moosebuild commented:
Job [Python 3.10 (Apptainer)](https://civet.inl.gov/job/1559680/) on fc235cf : invalidated by @GiudGiud

unrelared

moosebuild commented:
Job [Python 3.9](https://civet.inl.gov/job/1559683/) on fc235cf : invalidated by @GiudGiud

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1559973/) on fb29646 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/24542/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1560014/) on fb29646 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>447dca</th>
      <th colspan=4>#24542 fb2964</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.36%</td>
      <td>85.34%</td>
      <td>-0.02%</td>
      <td>33.33%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>88359</td>
      <td>88367</td>
      <td>+8</td>
      <td>12</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15150</td>
      <td>15178</td>
      <td>+28</td>
      <td>24</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24542/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24542/coverage/framework)


## Modules coverage

### Thermal hydraulics

<table>
  <thead>
    <tr>
      <th></th>
      <th>447dca</th>
      <th colspan=4>#24542 fb2964</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>90.39%</td>
      <td>90.37%</td>
      <td>-0.02%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>12553</td>
      <td>12530</td>
      <td>-23</td>
      <td>8</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1335</td>
      <td>1335</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24542/coverage_diff/thermal_hydraulics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24542/coverage/thermal_hydraulics)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/24542/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/24542/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/24542/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/24542/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/24542/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/24542/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/24542/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/24542/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/24542/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/24542/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/24542/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/24542/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/24542/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/24542/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/24542/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/24542/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/24542/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/24542/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/24542/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/24542/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/24542/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/24542/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/24542/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/24542/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/24542/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/24542/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/24542/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/24542/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 33.33% is less than the suggested 90.0%



This comment will be updated on new commits.

lindsayad reviewed (changes requested):
With the ICs in the framework now, we should test them there

joshuahansel commented on framework/src/ics/ScalarSolutionInitialCondition.C:
```diff
@@ -21,6 +21,10 @@ ScalarSolutionInitialCondition::validParams()
       "from_variable", "The name of the variable in the file that is to be extracted");
+  params.addClassDescription(
+      "Sets the initial condition from a scalar variable stored in an Exodus file, "
+      "retrieve by a SolutionUserObject");
```
```suggestion
      "retrieved by a SolutionUserObject");
```

joshuahansel commented on framework/src/ics/SolutionInitialCondition.C:
```diff
@@ -21,6 +21,9 @@ SolutionInitialCondition::validParams()
       "from_variable", "The name of the variable in the file that is to be extracted");
+  params.addClassDescription(
+      "Sets the initial condition from a field variable stored in an Exodus file, "
+      "retrieve by a SolutionUserObject");
```
```suggestion
      "retrieved by a SolutionUserObject");
```

joshuahansel commented on modules/thermal_hydraulics/doc/content/source/ics/RhoEAFromPressureTemperatureFunctionVelocityIC.md:
```diff
@@ -2,10 +2,11 @@
 
 This object computes the value of $rhoEA$ given pressure and temperature as variables and velocity as a function.
 
+!alert note
+This initial condition is typically already provided by the [flow model setup](syntax/FlowModelSetup/index.md).
+
```
```suggestion
```

joshuahansel replied:
See my comment on the linked file. `FlowModelSetup` isn't anything typically used. `FlowModelSinglePhase` doesn't use it at all, for instance (probably dumb).

joshuahansel commented on modules/thermal_hydraulics/doc/content/source/ics/RhoEAFromPressureTemperatureVelocityIC.md:
```diff
@@ -1,18 +1,9 @@
-!! Replace these lines with information regarding the RhoEAFromPressureTemperatureVelocityIC object.
-
-## Example Input File Syntax
+!alert note
+This initial condition is typically already provided by the [flow model setup](syntax/FlowModelSetup/index.md).
 
```
```suggestion
```

joshuahansel commented on modules/thermal_hydraulics/doc/content/source/ics/SpecificInternalEnergyIC.md:
```diff
@@ -1,20 +1,15 @@
+equal to one.
 
-!! Describe and include an example of how to use the SpecificInternalEnergyIC object.
+!alert note
+This initial condition is typically already provided by the [flow model setup](syntax/FlowModelSetup/index.md).
 
```
```suggestion
```

joshuahansel commented on modules/thermal_hydraulics/doc/content/source/ics/SpecificTotalEnthalpyIC.md:
```diff
@@ -1,20 +1,15 @@
+equal to one.
 
-!! Describe and include an example of how to use the SpecificTotalEnthalpyIC object.
+!alert note
+This initial condition is typically already provided by the [flow model setup](syntax/FlowModelSetup/index.md).
 
```
```suggestion
```

joshuahansel commented on modules/thermal_hydraulics/doc/content/source/ics/SumIC.md:
```diff
@@ -1,20 +1,9 @@
-## Example Input File Syntax
-
-!! Describe and include an example of how to use the SumIC object.
+The value of the other variables in the sum is naturally equal to their initial conditions,
+as initial conditions are only run once at the beginning of the simulation.
```
I don't understand what point you're making / why you're clarifying this.

GiudGiud replied:
just that things are as you'd want eg, this IC is ran last. Maybe it s not needed

joshuahansel commented on modules/thermal_hydraulics/doc/content/source/ics/VectorVelocityIC.md:
```diff
@@ -6,13 +6,11 @@ It computes:
 u_i = u_d d_i
 \end{equation}
-where $u$ is the velocity vector, $u_d$ is user-provided function and $d$ is the direction
+where $u$ is the velocity vector, $u_d$ is a user-provided scalar function and $d$ is the direction
 of the element where this IC is being applied.
```
```suggestion
where $u$ is the velocity vector, $\mathbf{d}$ is the unit direction vector
of the element where this IC is being applied,
and $u_d$ is a user-provided function for the component of velocity in the direction $\mathbf{d}$.
```

GiudGiud commented on modules/thermal_hydraulics/doc/content/syntax/FlowModelSetup/index.md:
```diff
@@ -0,0 +1,15 @@
+# FlowModelSetup System
```
This was kind of a place holder.

Good to know. I ll adapt the docs. I thought it was a core thing

GiudGiud replied:
Removed the page.

I still dont see any of these ICs in the THM tests so I suspect THM sets them programmatically

GiudGiud committed:
Address Josh's review

- remove note pointing to a cool class for initialization that is not actually in use

Co-authored-by: joshuahansel <joshua.hansel@inl.gov>

GiudGiud committed:
Rename solution ICs to follow MOOSE convention of IC for object names

GiudGiud committed:
Rework block restrict checks to be able to handle:
- no block names in source exodus
- no block restriction specified in IC block (ANY_BLOCK name)

GiudGiud committed:
Add test for solution IC and scalar IC, refs #24581

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1565889/) on 0cf7040 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24542/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24542/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a48e4b035578b9dc8f668e1c247b501ca2e5b75e`

GiudGiud commented:
Good call on the testing, this object wasnt working right for the regular Joe's mesh file

moosebuild commented:
Job [Conda (ARM Mac)](https://civet.inl.gov/job/1567569/) on ee32322 : invalidated by @GiudGiud

unrelated

GiudGiud committed:
Add a note to each IC that is usually not created by a user
The flow model creates them in the simulation based on component parameters
refs #15968 #19737

lindsayad commented on framework/src/ics/SolutionIC.C:
```diff
@@ -0,0 +1,83 @@
+  if (_solution_object.getSolutionFileType() == 1)
+  {
+    // remap block names this IC is defined on into the ExodusII file block IDs
+    const std::map<SubdomainName, SubdomainID> & block_names_to_ids_from =
```
```suggestion
    const auto & block_names_to_ids_from =
```

lindsayad commented on framework/src/ics/SolutionIC.C:
```diff
@@ -0,0 +1,83 @@
+
+    for (auto & blk_name : blocks_to_check)
+    {
+      auto jt = block_names_to_ids_from.find(blk_name);
```
```suggestion
      auto it = block_names_to_ids_from.find(blk_name);
```

lindsayad commented on framework/src/ics/SolutionIC.C:
```diff
@@ -0,0 +1,83 @@
+    for (auto & blk_name : blocks_to_check)
+    {
+      auto jt = block_names_to_ids_from.find(blk_name);
+      if (jt != block_names_to_ids_from.end())
+        _exo_block_ids.insert(jt->second);
```
```suggestion
      if (it != block_names_to_ids_from.end())
        _exo_block_ids.insert(it->second);
```

lindsayad commented on framework/src/ics/SolutionIC.C:
```diff
@@ -0,0 +1,83 @@
+        auto blk_id = _sys.mesh().getSubdomainID(blk_name);
+        // use the ids, it may be that the source file does not have block names
+        // and that we are using the id as the block name (block = 0 for example)
+        const std::map<SubdomainID, SubdomainName> & block_ids_to_names_from =
```
```suggestion
        const auto & block_ids_to_names_from =
```

lindsayad commented on framework/src/ics/SolutionIC.C:
```diff
@@ -0,0 +1,83 @@
+        const std::map<SubdomainID, SubdomainName> & block_ids_to_names_from =
+            _solution_object.getBlockIdsToNames();
+        if (block_ids_to_names_from.find(blk_id) != block_ids_to_names_from.end() &&
+            block_ids_to_names_from.find(blk_id)->second == "")
```
```suggestion
            block_ids_to_names_from.find(blk_id)->second.empty())
```

lindsayad commented on framework/src/ics/SolutionIC.C:
```diff
@@ -0,0 +1,83 @@
+            _solution_object.getBlockIdsToNames();
+        if (block_ids_to_names_from.find(blk_id) != block_ids_to_names_from.end() &&
+            block_ids_to_names_from.find(blk_id)->second == "")
+          _exo_block_ids.insert(jt->second);
```
```suggestion
          _exo_block_ids.insert(it->second);
```

lindsayad replied:
Wait this doesn't make any sense to me ... I thought that we didn't find anything ... I don't trust this iterator dereference

lindsayad replied:
Yea this line of code doesn't get hit in the coverage report

GiudGiud replied:
oh it s the wrong iterator

lindsayad reviewed (changes requested):
Have to run so review is incomplete

lindsayad commented on modules/thermal_hydraulics/doc/content/source/ics/SpecificTotalEnthalpyIC.md:
```diff
@@ -1,20 +1,16 @@
 
-!! Describe and include an example of how to use the SpecificTotalEnthalpyIC object.
+!alert note
+This initial condition is usually added to the `Simulation` by the `flow model`, based on the parameters
```
I'm not used to seeing code formatting with a space, `flow model`

lindsayad replied:
should it just be "flow model", no code formatting, or `flow_model`?

GiudGiud replied:
it is code but it is also code that is under the hood and undocumented
so it could be `FlowModel` but then if you search for that in the documentation you cant find it yet..

I think I ll just try to have that page ready in the near future

lindsayad commented on modules/thermal_hydraulics/doc/content/source/ics/SpecificVolumeIC.md:
```diff
@@ -1,20 +1,14 @@
 
-!! Describe and include an example of how to use the SpecificVolumeIC object.
+!alert note
+This initial condition is usually added to the `Simulation` by the `flow model`, depending on the parameters
```
same question as above

lindsayad commented on test/tests/ics/solution_ic/tests:
```diff
@@ -0,0 +1,27 @@
+      input = 'solution_ic.i'
+      cli_args = "Mesh/gmg/subdomain_ids='0 0 0 2' ICs/initial_cond_nl/block=2"
+      expect_err = "Block '2' does not exist in the file"
+      detail = "a solution is to be loaded at a block that does not exist in the source file"
```
```suggestion
      detail = "a solution is to be loaded at a block that does not exist in the source file."
```

GiudGiud committed:
Address review from Alex

- fix logic for unnamed subdomain
- use string for enum comparison
- refer to FlowModel in the ICs

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1570953/) on 7cc1152 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24542/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24542/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 9eb089dd6184886878b6d7fd6a2bf081fb29516e`

lindsayad reviewed (approved):
Great!
